### PR TITLE
💄 Add NFT Book gem & Add NFT type filter

### PR DESCRIPTION
**Pull request by AuroraHuang22**
截圖 2023-03-03 下午5 38 22
截圖 2023-03-03 下午8 54 24

https://user-images.githubusercontent.com/75730405/222681721-ac110936-5d0d-4c3f-87ba-c65c8c601b23.mp4

### CODE REVIEW

File: src/components/NFTGem/Wrapper.vue [hunk 1]
```diff
@@ -26,14 +26,20 @@ export default {
...
+    isNftBook: {
+      type: Boolean,
+      default: false,
+    },
   },
   computed: {
     ...mapGetters(['getNFTClassGemLevel']),
     gemName() {
-      return NFT_GEM_NAME[this.gemLevel];
+      return this.gemLevel === 'book'
```
Comment (nwingt) [inline src/components/NFTGem/Wrapper.vue:37]: I think 'book' should be defined as const

File: src/components/NFTGem/Wrapper.vue [hunk 2]
```diff
@@ -26,14 +26,20 @@ export default {
...
+        ? 'NFT-Book'
```
Comment (nwingt) [inline src/components/NFTGem/Wrapper.vue:38]: Should be localized?

File: src/components/NFTGem/Wrapper.vue [hunk 3]
```diff
@@ -84,7 +91,7 @@ export default {
           return 'hover:bg-[#ffd748]';
         case gemLevel <= 14:
           return 'hover:bg-[#ff6464]';
-        case gemLevel <= 15:
+        case gemLevel <= 16:
```
Comment (nwingt) [inline src/components/NFTGem/Wrapper.vue:94]: ??
```suggestion
        case this.gemLevel === 'book':
```

File: src/util/nft.js [hunk 4]
```diff
@@ -29,6 +29,13 @@ export const NFT_CLASS_LIST_SORTING_ORDER = {
   DESC: 'DESC',
 };
 
+export const NFT_TYPE_OPTIONS = {
+  ALL: 'All NFTs',
+  WNFT: 'Writing NFT',
+  BOOK: 'NFT Book',
+  OTHER: 'Other NFT',
```
Comment (nwingt) [inline src/util/nft.js:36]: Are these shown directly in UI, I prefer localized them

File: src/mixins/portfolio.js [hunk 5]
```diff
@@ -88,27 +92,126 @@ export const createPorfolioMixin = ({
...
+    validNftTypeOptions() {
+      const currentClassList =
+        this.currentTab === tabOptions.collected
+          ? this.nftClassListOfCollected
+          : this.nftClassListOfCreated;
+
+      const filteredOptions = Object.entries(NFT_TYPE_OPTIONS)
+        .filter(([key]) => {
+          switch (key) {
+            case 'WNFT':
```
Comment (nwingt) [inline src/mixins/portfolio.js:133]: Can we use `NFT_TYPE_OPTIONS.WNFT`?

File: src/mixins/portfolio.js [hunk 6]
```diff
@@ -405,8 +508,10 @@ export const createPorfolioMixin = ({
     changeTab(tab) {
       if (!tabOptions[tab]) return;
       this.syncRouteForTab(tab);
+      this.nftTypeFilter = NFT_TYPE_OPTIONS.ALL;
```
Comment (nwingt) [inline src/mixins/portfolio.js:511]: Seems more intuitive
```suggestion
      this.nftTypeFilter = NFT_TYPE_FITLER_OPTIONS.ALL;
```

File: src/pages/dashboard/index.vue [hunk 7]
```diff
@@ -51,13 +51,16 @@
         :portfolio-items-sorting="currentNFTClassListSorting"
         :portfolio-items-sorting-order="currentNFTClassListSortingOrder"
         :portfolio-items-sorting-option-list="currentNFTClassSortingOptionList"
-        :portfolio-items-filtering="{ creator: nftCreatorFilter }"
+        :portfolio-items-creator-filtering="{ creator: nftCreatorFilter }"
```
Comment (nwingt) [inline src/pages/dashboard/index.vue:54]: Why object?

File: src/util/nft.js [hunk 8]
```diff
@@ -29,6 +29,13 @@ export const NFT_CLASS_LIST_SORTING_ORDER = {
   DESC: 'DESC',
 };
 
+export const NFT_TYPE_OPTIONS = {
```
Comment (nwingt) [inline src/util/nft.js:32]:
```suggestion
export const NFT_TYPE_FILTER_OPTIONS = {
```

File: src/mixins/portfolio.js [hunk 9]
```diff
@@ -88,27 +92,126 @@ export const createPorfolioMixin = ({
...
+          const isLikecoinChainNFTBook = checkIsNftBook(classMetadata);
+          return isLikecoinChainNFTBook;
+        })
+        .reduce((map, nft) => map.set(nft.classId, nft), new Map());
+      return nftClassMapOfNftBook;
+    },
     nftClassListOfOther() {
       return [...this.nftClassMapOfOther.values()];
     },
+    validNftTypeOptions() {
```
Comment (nwingt) [inline src/mixins/portfolio.js:124]:
```suggestion
    nftTypeFilteringOptions() {
```

File: src/components/NFTPortfolio/MainView.vue [hunk 10]
```diff
@@ -260,7 +295,15 @@ export default {
...
-    portfolioItemsFiltering: {
+    portfolioItemsCreatorFiltering: {
+      type: Array,
+      default: () => [],
+    },
+    portfolioItemsTypeFiltering: {
+      type: String,
+      required: true,
+    },
+    portfolioItemsTypeOptions: {
```
Comment (nwingt) [inline src/components/NFTPortfolio/MainView.vue:306]:
```suggestion
    portfolioItemsTypeFilteringOptions: {
```

File: src/components/NFTPortfolio/MainView.vue [hunk 11]
```diff
@@ -323,6 +366,22 @@ export default {
...
+    creatorFilterLabelText() {
+      const isSelected = this.portfolioCollectedCreatorListWithSorting.filter(
+        creator => creator.isSelected
+      );
+      return isSelected.length
+        ? this.$t('filter_menu_creator_selected', { num: isSelected.length })
+        : this.$t('filter_menu_creator');
+    },
+    nftTypeOptions() {
+      return Object.values(this.portfolioItemsTypeOptions);
```
Comment (nwingt) [inline src/components/NFTPortfolio/MainView.vue:378]: I think this list of options should return translated labels
```suggestion
      return Object.values(this.portfolioItemsTypeOptions).map((value) => ({
        value,
        label: this.getPortfolioTypeFilteringLabel(value),
      });
```

**Comment by nwingt** [comment]
I think this feature can be rebased to master?

File: src/constant/index.js [hunk 12]
```diff
@@ -162,6 +162,10 @@ export const NFT_GEM_NAME = [
   'Diamond',
 ];
 
+export const NFT_BOOK = 'NFT_BOOK';
+
+export const NFT_GEM_NAME_BOOK = 'NFT-Book';
```
[the remark this answers is not in the document]
Comment (AuroraHuang22) [reply src/constant/index.js:167]: yes, the [gem name](https://github.com/likecoin/liker-land/pull/979/files#diff-1eb6fd5fdfc5858c6d9b5829d6fb4eb0656d0184e9c347292796a40a3feff37bR38) will be shown as tooltip when hovering

File: src/mixins/portfolio.js [hunk 13]
```diff
@@ -405,8 +508,10 @@ export const createPorfolioMixin = ({
...
       if (!tabOptions[tab]) return;
       this.syncRouteForTab(tab);
+      this.nftTypeFilter = NFT_TYPE_FILTER_OPTIONS.ALL;
+      this.nftCreatorFilter = [];
     },
-    handleNFTClassListFilteringChange({ type, value }) {
+    handleNFTClassListCreatorChange({ type, value }) {
       logTrackerEvent(
         this,
         'portfolio',
```
Comment (williamchong) [inline src/mixins/portfolio.js:517]: Since the type always be `creator` now, we just just omit the `type` and hardcode `creator`

**Review by williamchong — CHANGES_REQUESTED** [review]

File: src/mixins/portfolio.js [hunk 14]
```diff
@@ -88,27 +92,126 @@ export const createPorfolioMixin = ({
...
+            ({ classId }) => this.nftClassMapOfOther.has(classId)
+          );
+          break;
+
+        default:
+          break;
+      }
+      return filteredCollectedList;
     },
     nftClassListOfCreatedExcludedOther() {
```
[the remark this answers is not in the document]
Comment (nwingt) [reply src/mixins/portfolio.js:186]: Agree

File: src/mixins/portfolio.js [hunk 15]
```diff
@@ -88,27 +92,126 @@ export const createPorfolioMixin = ({
...
+            case NFT_TYPE_FILTER_OPTIONS.ALL:
+              return true;
+            default:
+              return false;
+          }
+        })
+        .reduce((obj, [key, value]) => ({ ...obj, [key]: value }), {});
+      return filteredOptions;
+    },
     nftClassListOfCollectedExcludedOther() {
```
[the remark this answers is not in the document]
Comment (nwingt) [reply src/mixins/portfolio.js:156]: Agree

File: src/mixins/portfolio.js [hunk 16]
```diff
@@ -88,27 +92,126 @@ export const createPorfolioMixin = ({
...
+        .filter(nft => {
+          const classMetadata = this.getNFTClassMetadataById(nft.classId);
+          const isWritingNFT = checkIsWritingNFT(classMetadata);
+          const isLikecoinChainNFTBook = checkIsNftBook(classMetadata);
+          return !isWritingNFT && !isLikecoinChainNFTBook;
+        })
+        .reduce((map, nft) => map.set(nft.classId, nft), new Map());
       return nftClassMapOfOther;
     },
+    nftClassMapOfNftBook() {
```
Comment (nwingt) [inline src/mixins/portfolio.js:105]:
```suggestion
    nftClassMapOfNFTBook() {
```

File: src/util/nft.js [hunk 17]
```diff
@@ -202,6 +209,12 @@ export function checkIsWritingNFT(classMetadata) {
   );
 }
 
+export function checkIsNftBook(classMetadata) {
```
Comment (nwingt) [inline src/util/nft.js:212]:
```suggestion
export function checkIsNFTBook(classMetadata) {
```

**Review by nwingt — COMMENTED** [review]
I saw some BookNFT and NFTBook, it should be latter

File: src/components/NFTPortfolio/MainView.vue [hunk 18]
```diff
@@ -353,13 +414,6 @@ export default {
         },
       ];
 
-      if (this.isShowOtherTab) {
-        items.push({
-          text: this.$t('nft_portfolio_page_label_other'),
-          value: tabOptions.other,
-        });
-      }
-
```
[the remark this answers is not in the document]
Comment (AuroraHuang22) [reply src/components/NFTPortfolio/MainView.vue:362]: yes

File: src/mixins/portfolio.js [hunk 19]
```diff
@@ -84,12 +84,22 @@ export const createPortfolioMixin = ({
...
     nftClassMapOfOther() {
       const nftClassMapOfOther = Array.from(this.allNFTClassMap.values())
         .filter(nft => {
           const classMetadata = this.getNFTClassMetadataById(nft.classId);
           const isWritingNFT = checkIsWritingNFT(classMetadata);
-          const isLikecoinChainNFTBook = checkIsNftBook(classMetadata);
+          const isLikecoinChainNFTBook = checkIsNFTBook(classMetadata);
           return !isWritingNFT && !isLikecoinChainNFTBook;
         })
         .reduce((map, nft) => map.set(nft.classId, nft), new Map());
```
Comment (williamchong) [inline src/mixins/portfolio.js:105]: We can build `nftClassMapOfOther` using `nftClassMapOfWritingNft` and `nftClassMapOfNFTBook`

**Review by williamchong — CHANGES_REQUESTED** [review]

File: src/locales/en.json [hunk 20]
```diff
@@ -181,8 +181,10 @@
   "FeaturedPage": {
     "headerLabel": "Today"
   },
-  "filter_menu_creator": "Filter",
+  "filter_menu_creator": "All Creators",
+  "filter_menu_creator_selected": "{num} Creator(s)",
```
Comment (nwingt) [inline src/locales/en.json:185]: Can use `$tc()`
Ref: https://kazupon.github.io/vue-i18n/guide/pluralization.html#pluralization
```suggestion
  "filter_menu_creator_selected": "{num} Creator | {num} Creators",
```

**Review by nwingt — APPROVED** [review]

**Pull request merged**